swan-lake/development-tutorials/test-document-the-code/ballerina-profiler.md

```diff
@@ -0,0 +1,365 @@
+
+```ballerina
+import ballerina/io;
+import ballerina/random;
+
+public function main() returns error? {
+    int[] arr = check createRandomIntArray(check float:pow(10, 2).cloneWithType(int));
+    int[] sortedArr = bubbleSort(arr);
+    boolean isSorted = isSortedArray(sortedArr);
+    io:println("Is the array sorted? " + isSorted.toString());
+}
+
+public isolated function bubbleSort(int[] arr) returns int[] {
+    int n = arr.length();
+    int temp = 0;
+    boolean swapped = false;
+    foreach int i in 0 ... n - 2 {
+        foreach int j in 1 ... n - 1 - i {
+            if (arr[j - 1] > arr[j]) {
+                temp = arr[j - 1];
+                arr[j - 1] = arr[j];
+                arr[j] = temp;
+                swapped = true;
+            }
+        }
+        if (!swapped) {
+            break;
+        }
+    }
+    return arr;
+}
+
+isolated function isSortedArray(int[] sortedArr) returns boolean {
+    foreach int i in 0 ..< sortedArr.length() - 1 {
+        if (sortedArr[i] > sortedArr[i + 1]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+isolated function createRandomIntArray(int size) returns int[]|error {
+    int[] array = [];
+    foreach int i in 0 ..< size {
+        array.push(check random:createIntInRange(0, int:MAX_VALUE));
+    }
+    return array;
+}
+```
+
+Run and profile the Ballerina package using the `bal profile` CLI command.
+
+```
+$ bal profile
+```
+
+This will give the following output.
+
+```
+Compiling source
+        profiler_demo/sort:0.1.0
+
+Generating executable
+        target/bin/sort.jar
+
+================================================================================
+Ballerina Profiler: Profiling...
+================================================================================
+Note: This is an experimental feature, which supports only a limited set of functionality.
+[1/6] Initializing...
+[2/6] Copying executable...
+[3/6] Performing analysis...
+[4/6] Instrumenting functions...
+ ○ Instrumented module count: 31
+ ○ Instrumented function count: 1016
+[5/6] Running executable...
+Is the array sorted? true
+[6/6] Generating output...
+ ○ Execution time: 3 seconds 
+ ○ Output: target/bin/ProfilerOutput.html
+-------------------------------------------------------------------------------
+```
+
+Open the `target/bin/ProfilerOutput.html` file using a web browser window to examine the Ballerina Profiler output to find the slow-running functions and performance bottlenecks, which can then be addressed to improve the program’s overall performance.
+
+## Profiler output and available details
+
+![Profiler output](/learn/images/profiler-output.png "Profiler output")
+
+The flame graph generated by the Ballerina Profiler contains the following details.
+
+1. Function call stack
+
+2. Time taken to execute each function and its percentage
+
+Flame graph also has the following features.
```

Comment:
```suggestion
Flame graph also has the following features.
```
Do the steps below demonstrate one feature or multiple?

Reply:
Multiple features: numbered as 1 2 3 4